src/vmm/src/resources.rs

```diff
@@ -539,4 +653,240 @@ mod tests {
+        let expected_file_path = tmp_file.as_path().to_path_buf();
+        let block_cfg_to_be_updated = vm_resources.block.config_list.get(0).unwrap().clone();
+        assert_ne!(block_cfg_to_be_updated.path_on_host, expected_file_path);
+        vm_resources
+            .update_block_device_path(
+                block_cfg_to_be_updated.drive_id.clone(),
+                String::from(expected_file_path.to_str().unwrap()),
+            )
+            .unwrap();
+        let updated_block_cfg = vm_resources.block.config_list.get(0).unwrap();
+        assert_eq!(updated_block_cfg.path_on_host, expected_file_path);
+
+        // InvalidBlockDeviceId
+        assert_eq!(
+            vm_resources.update_block_device_path("id_does_not_exist".to_string(), "".to_string(),),
+            Err(DriveError::InvalidBlockDeviceID)
+        );
+
+        // CannotOpenBlockDevice
+        assert!(vm_resources
+            .update_block_device_path(
+                block_cfg_to_be_updated.drive_id.clone(),
+                "/does/not/exist".to_string(),
+            )
+            .is_err());
+    }
+
+    #[test]
+    fn test_update_net_rate_limiters() {
+        let bw_tb = TokenBucketConfig {
+            size: 15,
+            one_time_burst: Some(5),
+            refill_time: 5,
+        };
+
+        let ops_tb = TokenBucketConfig {
+            size: 10,
+            one_time_burst: Some(2),
+            refill_time: 2,
+        };
+
+        let expected_rl_cfg = RateLimiterConfig {
+            bandwidth: Some(bw_tb),
+            ops: Some(ops_tb),
+        };
+
+        let mut vm_resources = default_vm_resources();
+        let actual_net_cfg = vm_resources.network_interface.iter().next().unwrap();
+        let actual_rx_rl_cfg = actual_net_cfg.rx_rate_limiter.unwrap();
+        let actual_tx_rl_cfg = actual_net_cfg.tx_rate_limiter.unwrap();
+        assert_ne!(actual_rx_rl_cfg, expected_rl_cfg);
+        assert_ne!(actual_tx_rl_cfg, expected_rl_cfg);
+
+        let mut net_if_cfg_update = NetworkInterfaceUpdateConfig {
+            iface_id: actual_net_cfg.iface_id.clone(),
+            rx_rate_limiter: Some(expected_rl_cfg),
+            tx_rate_limiter: Some(expected_rl_cfg),
+        };
+        vm_resources
+            .update_net_rate_limiters(net_if_cfg_update.clone())
+            .unwrap();
+        let actual_net_cfg = vm_resources.network_interface.iter().next().unwrap();
+        let actual_rx_rl_cfg = actual_net_cfg.rx_rate_limiter.unwrap();
+        let actual_tx_rl_cfg = actual_net_cfg.tx_rate_limiter.unwrap();
+        assert_eq!(actual_rx_rl_cfg, expected_rl_cfg);
+        assert_eq!(actual_tx_rl_cfg, expected_rl_cfg);
+
+        // DeviceIdNotFound
+        net_if_cfg_update.iface_id = "net_if_does_not_exist".to_string();
+        assert_eq!(
+            vm_resources.update_net_rate_limiters(net_if_cfg_update),
+            Err(NetworkInterfaceError::DeviceIdNotFound)
+        );
```

Comment:
For checking the `DeviceIdNotFound` error, you can replace this `assert_eq()` with:
 `match vm_resources.update_net_rate_limiters(net_if_cfg_update) {
            Err(NetworkInterfaceError::DeviceIdNotFound { .. }) => (),
            _ => unreachable!(),
        }`
This way you don't have to implement `PartialEq` for `NetworkInterfaceError`.